crates/flightrepl/src/completer.rs

```diff
@@ -0,0 +1,502 @@
+use crate::EditorHelper;
+use arrow_flight::flight_service_client::FlightServiceClient;
+use datafusion::arrow::array::{Array, StringArray};
+use rustyline::Context;
+use rustyline::completion::{Completer, Pair};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::{RwLock, oneshot};
+use tokio::time::interval;
+use tonic::transport::Channel;
+
+#[rustfmt::skip]
+const SQL_KEYWORDS: &[&str] = &[
+    // Core SQL keywords
+    "select", "from", "where", "and", "or", "not", "in", "like", "is", "null",
+    "true", "false", "asc", "desc", "limit", "offset", "order", "by", "group",
+    "having", "as", "distinct", "case", "when", "then", "else", "end",
+
+    // JOIN operations
+    "join", "inner", "left", "right", "full", "outer", "cross", "natural", "using", "on",
+
+    // DML operations
+    "insert", "into", "values", "update", "set", "delete",
+
+    // DDL operations
+    "create", "table", "alter", "drop", "index", "view", "database", "schema",
+];
+
+#[rustfmt::skip]
+const FUNCTION_NAMES: &[&str] = &[
```

Comment:
Keeping parity with this const as DataFusion and Spice changes could be tough. It would be cool if the autocomplete could ask Spice via query what UDFs it supports:

- The `SessionContext` function registry can give you a list of supported names: [`udfs()`](https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionContext.html#method.udfs)
- You could expose a table such as `spice_functions` as a UDTF: [(simple udtf example)](https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/simple_udtf.rs).

Then on REPL boot you could obtain this list via a query, and never have to worry about updating anything!

Reply:
One approach here to keep at least the names in sync (between runtime and flightrepl) is:
1. Define a function in `runtime` crate that provides all runtime-specific UDFs name (e.g. `available_runtime_udf_names`). 
2. Provide to flightrepl via `spiced` binary, [here](https://github.com/spiceai/spiceai/blob/b063058d74960c5f9caff38312d5144625e12df8/bin/spiced/src/main.rs#L77)
    ```rust
    if let Err(e) = tokio_runtime.block_on(flightrepl::run(args.repl_config), runtime::available_runtime_udf_names()) {
    ```

Reply:
And then make sure to have comments on the relevant `runtime` functions so that when other people add UDFs, they know to update `available_runtime_udf_names`. We do similar [here](https://github.com/spiceai/spiceai/blob/b063058d74960c5f9caff38312d5144625e12df8/crates/runtime/src/datafusion/udf.rs#L22).

Reply:
Awesome! Thanks, will do!